### PR TITLE
    MGMT-12214: extract reconciler logic for using HASC

### PR DESCRIPTION
**Pull request by danielerez**
Extracted the following from ASC reconciler for using in HASC:
* Finalizers
* image-service StatefulSet
* Storage validation
* IPXE HTTPRoute

## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/hypershiftagentserviceconfig_controller.go [hunk 1]
```diff
@@ -141,17 +211,23 @@ func (hr *HypershiftAgentServiceConfigReconciler) Reconcile(origCtx context.Cont
 			return result, err
 		}
 	}
+	return ctrl.Result{}, nil
+}
+
+func (hr *HypershiftAgentServiceConfigReconciler) reconcileSpokeComponents(ctx context.Context, log *logrus.Entry, asc ASC) (ctrl.Result, error) {
+	spokeComponents := []component{}
```
Comment (slaviered) [inline internal/controller/controllers/hypershiftagentserviceconfig_controller.go:218]: service account should be part of the RBAC array. It has the same semantics. also is order important here?

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 2]
```diff
@@ -223,108 +223,34 @@ func (r *AgentServiceConfigReconciler) Reconcile(origCtx context.Context, req ct
...
-			return ctrl.Result{Requeue: true}, err
-		}
-
-		controllerutil.RemoveFinalizer(instance, agentServiceConfigFinalizerName)
-		if err := r.Update(ctx, instance); err != nil {
-			log.WithError(err).Error("failed to remove finalizer from AgentServiceConfig")
-			return ctrl.Result{Requeue: true}, err
-		}
-
+	// Ensure relevant finalizers exist (cleanup on deletion)
```
Comment (slaviered) [inline internal/controller/controllers/agentserviceconfig_controller.go:226]: The order here is important: first you need to check the deletion time and then ensure.

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 3]
```diff
@@ -399,6 +325,16 @@ func getComponents() []component {
 		{"WebHookClusterRoleBinding", aiv1beta1.ReasonWebHookClusterRoleBindingFailure, newWebHookClusterRoleBinding},
 		{"WebHookAPIService", aiv1beta1.ReasonWebHookAPIServiceFailure, newWebHookAPIService},
 	}
+
+	// Additional routes need to be synced if HTTP iPXE routes are exposed
+	if exposeIPXEHTTPRoute(spec) {
```
Comment (slaviered) [inline internal/controller/controllers/agentserviceconfig_controller.go:330]: where is the delete part?

File: internal/controller/controllers/hypershiftagentserviceconfig_controller.go [hunk 4]
```diff
@@ -126,31 +134,59 @@ func (hr *HypershiftAgentServiceConfigReconciler) Reconcile(origCtx context.Cont
...
+	valid, err = validate(ctx, log, asc)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	if !valid {
+		return ctrl.Result{}, nil
+	}
+
+	// Remove IPXE HTTP routes if not needed
+	if err = cleanHTTPRoute(ctx, log, asc); err != nil {
```
Comment (slaviered) [inline internal/controller/controllers/hypershiftagentserviceconfig_controller.go:163]: it's either or relationship wtih the configuration of the route as far as I understand.

File: internal/controller/controllers/hypershiftagentserviceconfig_controller.go [hunk 5]
```diff
@@ -126,31 +134,59 @@ func (hr *HypershiftAgentServiceConfigReconciler) Reconcile(origCtx context.Cont
...
+		return ctrl.Result{}, err
+	}
+
 	// Switch to spoke client in the reconciler context
 	// TODO: extract client from AgentServiceConfigReconcileContext to avoid sharing
 	//       it with both hub and spoke clients.
 	asc.rec.Client = spokeClient
 
 	// Ensure instance namespace exists on spoke cluster
-	if err := hr.ensureSpokeNamespace(ctx, log, asc); err != nil {
```
Comment (slaviered) [inline internal/controller/controllers/hypershiftagentserviceconfig_controller.go:148]: This should be done on the start. there is no point in continuing the reconciliation of there is no namespace on the spoke side. It should be a pre-condition for the entire process

File: internal/controller/controllers/agentserviceconfig_controller.go (hunk 3 above)
Comment (danielerez) [reply internal/controller/controllers/agentserviceconfig_controller.go:330]: In cleanHTTPRoute func.

File: internal/controller/controllers/hypershiftagentserviceconfig_controller.go (hunk 4 above)
Comment (danielerez) [reply internal/controller/controllers/hypershiftagentserviceconfig_controller.go:163]: Right, we check if need (exposeIPXEHTTPRoute) in the cleanHTTPRoute func.

File: internal/controller/controllers/hypershiftagentserviceconfig_controller.go (hunk 5 above)
Comment (danielerez) [reply internal/controller/controllers/hypershiftagentserviceconfig_controller.go:148]: The ensure func creates the namespace if missing, so it should be fine..

**Comment by danielerez** [comment]
/override ci/prow/edge-e2e-metal-assisted-capi
/override ci/prow/images
(not related to changes)

**Pull request merged**